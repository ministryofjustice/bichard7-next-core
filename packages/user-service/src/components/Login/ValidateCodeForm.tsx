import React from "react"
import Form from "components/Form"
import { RememberForm } from "components/Login/RememberForm"
import Paragraph from "components/Paragraph"
import TextInput from "components/TextInput"
import NotReceivedEmail from "components/NotReceivedEmail"
import Button from "components/Button"

interface ValidateCodeFormProps {
  csrfToken: string
  emailAddress: string | undefined
  validationCode: string | undefined
  invalidCodeError: string | undefined
  stageName: "loginStage" | "resetStage"
  stageValue: "validateCode"
  sendAgainUrl: string
  showRememberForm?: boolean
  RememberFormComponent?: React.FC<{ checked: boolean }>
  buttonText?: string
}

const ValidateCodeForm: React.FC<ValidateCodeFormProps> = ({
  csrfToken,
  emailAddress,
  validationCode,
  invalidCodeError,
  stageName,
  stageValue,
  sendAgainUrl,
  showRememberForm = false,
  buttonText
}) => {
  const textInputWidth = "30"
  const defaultButtonText = stageName === "loginStage" ? "Sign in" : "Next"
  const finalButtonText = buttonText || defaultButtonText

  return (
    <Form method="post" csrfToken={csrfToken}>
      <h1 className="govuk-heading-xl">{"Check your email"}</h1>
      <Paragraph>
        {"We have sent a code to:"} <b>{emailAddress}</b>
      </Paragraph>
      <Paragraph>{`Your code can take up to 5 minutes to arrive. Check your spam folder if you don't get an email.`}</Paragraph>
      <Paragraph className="govuk-!-padding-bottom-4">{`The code will expire after 30 minutes.`}</Paragraph>
      <input id="email" name="emailAddress" type="hidden" value={emailAddress} />
      <input type="hidden" name={stageName} value={stageValue} />
      <TextInput
        id="validationCode"
        name="validationCode"
        label="Security Code"
        labelSize="s"
        hint="Enter the security code"
        type="text"
        width={textInputWidth}
        value={validationCode}
        error={invalidCodeError}
        optionalProps={{ autocomplete: "off", "aria-autocomplete": "none" }}
      />
      {showRememberForm && RememberForm && <RememberForm checked={false} />}
      <Button noDoubleClick>{finalButtonText}</Button>
      <NotReceivedEmail sendAgainUrl={sendAgainUrl} />
    </Form>
  )
}

export default ValidateCodeForm
