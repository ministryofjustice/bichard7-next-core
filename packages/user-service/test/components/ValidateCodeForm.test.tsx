import { render } from "@testing-library/react"
import ValidateCodeForm from "components/Login/ValidateCodeForm"

describe("ValidateCodeForm", () => {
  const commonProps = {
    csrfToken: "mock-csrf-token",
    emailAddress: "bichard01@example.com",
    validationCode: "123456",
    validationCodeError: undefined,
    stageValue: "validateCode" as const,
    sendAgainUrl: "/send-code-again"
  }

  it("should render correctly for 'resetStage' (Next button) with no error", () => {
    const props = {
      ...commonProps,
      stageName: "resetStage" as const
    }
    const { container } = render(<ValidateCodeForm {...props} />)

    expect(container).toMatchSnapshot()
  })

  it("should render correctly for 'loginStage' (Sign in button) and show RememberForm", () => {
    const props = {
      ...commonProps,
      stageName: "loginStage" as const,
      showRememberForm: true
    }
    const { container } = render(<ValidateCodeForm {...props} />)

    expect(container).toMatchSnapshot()
  })

  it("should render correctly when a validationCodeError is present", () => {
    const props = {
      ...commonProps,
      stageName: "resetStage" as const,
      validationCodeError: "Incorrect security code"
    }
    const { container } = render(<ValidateCodeForm {...props} />)

    expect(container).toMatchSnapshot()
  })

  it("should render with custom button text if provided", () => {
    const props = {
      ...commonProps,
      stageName: "resetStage" as const,
      buttonText: "Verify Code"
    }
    const { container } = render(<ValidateCodeForm {...props} />)

    expect(container).toMatchSnapshot()
  })
})
