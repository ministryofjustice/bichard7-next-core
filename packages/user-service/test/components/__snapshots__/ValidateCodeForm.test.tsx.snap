// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ValidateCodeForm should render correctly for 'loginStage' (Sign in button) and show RememberForm 1`] = `
<div>
  <form
    method="post"
  >
    <input
      name="CSRFToken"
      type="hidden"
      value="mock-csrf-token"
    />
    <h1
      class="govuk-heading-xl"
    >
      Check your email
    </h1>
    <p
      class="govuk-body"
    >
      We have sent a code to:
       
      <b>
        bichard01@example.com
      </b>
    </p>
    <p
      class="govuk-body"
    >
      Your code can take up to 5 minutes to arrive. Check your spam folder if you don't get an email.
    </p>
    <p
      class="govuk-body govuk-!-padding-bottom-4"
    >
      The code will expire after 30 minutes
    </p>
    <input
      id="email"
      name="emailAddress"
      type="hidden"
      value="bichard01@example.com"
    />
    <input
      name="loginStage"
      type="hidden"
      value="validateCode"
    />
    <div
      class="govuk-form-group"
    >
      <label
        class="govuk-label govuk-label--s"
        for="validationCode"
      >
        Security Code
      </label>
      <div
        class="govuk-hint"
        id="validationCode-hint"
      >
        Enter the security code
      </div>
      <input
        aria-autocomplete="none"
        aria-describedby="validationCode-hint "
        autocomplete="off"
        class="govuk-input govuk-input--width-30"
        data-test="text-input_validationCode"
        id="validationCode"
        name="validationCode"
        type="text"
        value="123456"
      />
    </div>
    <button
      class="govuk-button"
      data-module="govuk-button"
      data-prevent-double-click="true"
      type="submit"
    >
      Sign in
    </button>
    <details
      class="govuk-details"
      data-module="govuk-details"
    >
      <summary
        class="govuk-details__summary"
        data-test="problem-with-code"
      >
        <span
          class="govuk-details__summary-text"
        >
          Problem with the code?
        </span>
      </summary>
      <div
        class="govuk-details__text"
      >
        <p>
          We can 
          <a
            class="govuk-link"
            data-test="send-code-again-link"
            href="/send-code-again?notYou=false&action=sendCodeAgain"
          >
            send the code again
          </a>
           or you can 
          <a
            class="govuk-link"
            data-test="not-you-link"
            href="/send-code-again?notYou=true"
          >
            use a different email address.
          </a>
        </p>
        <p>
          If you don't know your email address, contact the member of your team responsible for managing Bichard7 accounts.
        </p>
      </div>
    </details>
  </form>
</div>
`;

exports[`ValidateCodeForm should render correctly for 'resetStage' (Next button) with no error 1`] = `
<div>
  <form
    method="post"
  >
    <input
      name="CSRFToken"
      type="hidden"
      value="mock-csrf-token"
    />
    <h1
      class="govuk-heading-xl"
    >
      Check your email
    </h1>
    <p
      class="govuk-body"
    >
      We have sent a code to:
       
      <b>
        bichard01@example.com
      </b>
    </p>
    <p
      class="govuk-body"
    >
      Your code can take up to 5 minutes to arrive. Check your spam folder if you don't get an email.
    </p>
    <p
      class="govuk-body govuk-!-padding-bottom-4"
    >
      The code will expire after 30 minutes
    </p>
    <input
      id="email"
      name="emailAddress"
      type="hidden"
      value="bichard01@example.com"
    />
    <input
      name="resetStage"
      type="hidden"
      value="validateCode"
    />
    <div
      class="govuk-form-group"
    >
      <label
        class="govuk-label govuk-label--s"
        for="validationCode"
      >
        Security Code
      </label>
      <div
        class="govuk-hint"
        id="validationCode-hint"
      >
        Enter the security code
      </div>
      <input
        aria-autocomplete="none"
        aria-describedby="validationCode-hint "
        autocomplete="off"
        class="govuk-input govuk-input--width-30"
        data-test="text-input_validationCode"
        id="validationCode"
        name="validationCode"
        type="text"
        value="123456"
      />
    </div>
    <button
      class="govuk-button"
      data-module="govuk-button"
      data-prevent-double-click="true"
      type="submit"
    >
      Next
    </button>
    <details
      class="govuk-details"
      data-module="govuk-details"
    >
      <summary
        class="govuk-details__summary"
        data-test="problem-with-code"
      >
        <span
          class="govuk-details__summary-text"
        >
          Problem with the code?
        </span>
      </summary>
      <div
        class="govuk-details__text"
      >
        <p>
          We can 
          <a
            class="govuk-link"
            data-test="send-code-again-link"
            href="/send-code-again?notYou=false&action=sendCodeAgain"
          >
            send the code again
          </a>
           or you can 
          <a
            class="govuk-link"
            data-test="not-you-link"
            href="/send-code-again?notYou=true"
          >
            use a different email address.
          </a>
        </p>
        <p>
          If you don't know your email address, contact the member of your team responsible for managing Bichard7 accounts.
        </p>
      </div>
    </details>
  </form>
</div>
`;

exports[`ValidateCodeForm should render correctly when a validationCodeError is present 1`] = `
<div>
  <form
    method="post"
  >
    <input
      name="CSRFToken"
      type="hidden"
      value="mock-csrf-token"
    />
    <h1
      class="govuk-heading-xl"
    >
      Check your email
    </h1>
    <p
      class="govuk-body"
    >
      We have sent a code to:
       
      <b>
        bichard01@example.com
      </b>
    </p>
    <p
      class="govuk-body"
    >
      Your code can take up to 5 minutes to arrive. Check your spam folder if you don't get an email.
    </p>
    <p
      class="govuk-body govuk-!-padding-bottom-4"
    >
      The code will expire after 30 minutes
    </p>
    <input
      id="email"
      name="emailAddress"
      type="hidden"
      value="bichard01@example.com"
    />
    <input
      name="resetStage"
      type="hidden"
      value="validateCode"
    />
    <div
      class="govuk-form-group govuk-form-group--error"
    >
      <label
        class="govuk-label govuk-label--s"
        for="validationCode"
      >
        Security Code
      </label>
      <div
        class="govuk-hint"
        id="validationCode-hint"
      >
        Enter the security code
      </div>
      <span
        class="govuk-error-message"
        data-test="text-input_validationCode-error"
        id="validationCode-error"
      >
        <span
          class="govuk-visually-hidden"
        >
          Error:
        </span>
         
        Incorrect security code
      </span>
      <input
        aria-autocomplete="none"
        aria-describedby="validationCode-hint validationCode-error"
        autocomplete="off"
        class="govuk-input govuk-input--width-30 govuk-input--error"
        data-test="text-input_validationCode"
        id="validationCode"
        name="validationCode"
        type="text"
        value="123456"
      />
    </div>
    <button
      class="govuk-button"
      data-module="govuk-button"
      data-prevent-double-click="true"
      type="submit"
    >
      Next
    </button>
    <details
      class="govuk-details"
      data-module="govuk-details"
    >
      <summary
        class="govuk-details__summary"
        data-test="problem-with-code"
      >
        <span
          class="govuk-details__summary-text"
        >
          Problem with the code?
        </span>
      </summary>
      <div
        class="govuk-details__text"
      >
        <p>
          We can 
          <a
            class="govuk-link"
            data-test="send-code-again-link"
            href="/send-code-again?notYou=false&action=sendCodeAgain"
          >
            send the code again
          </a>
           or you can 
          <a
            class="govuk-link"
            data-test="not-you-link"
            href="/send-code-again?notYou=true"
          >
            use a different email address.
          </a>
        </p>
        <p>
          If you don't know your email address, contact the member of your team responsible for managing Bichard7 accounts.
        </p>
      </div>
    </details>
  </form>
</div>
`;

exports[`ValidateCodeForm should render with custom button text if provided 1`] = `
<div>
  <form
    method="post"
  >
    <input
      name="CSRFToken"
      type="hidden"
      value="mock-csrf-token"
    />
    <h1
      class="govuk-heading-xl"
    >
      Check your email
    </h1>
    <p
      class="govuk-body"
    >
      We have sent a code to:
       
      <b>
        bichard01@example.com
      </b>
    </p>
    <p
      class="govuk-body"
    >
      Your code can take up to 5 minutes to arrive. Check your spam folder if you don't get an email.
    </p>
    <p
      class="govuk-body govuk-!-padding-bottom-4"
    >
      The code will expire after 30 minutes
    </p>
    <input
      id="email"
      name="emailAddress"
      type="hidden"
      value="bichard01@example.com"
    />
    <input
      name="resetStage"
      type="hidden"
      value="validateCode"
    />
    <div
      class="govuk-form-group"
    >
      <label
        class="govuk-label govuk-label--s"
        for="validationCode"
      >
        Security Code
      </label>
      <div
        class="govuk-hint"
        id="validationCode-hint"
      >
        Enter the security code
      </div>
      <input
        aria-autocomplete="none"
        aria-describedby="validationCode-hint "
        autocomplete="off"
        class="govuk-input govuk-input--width-30"
        data-test="text-input_validationCode"
        id="validationCode"
        name="validationCode"
        type="text"
        value="123456"
      />
    </div>
    <button
      class="govuk-button"
      data-module="govuk-button"
      data-prevent-double-click="true"
      type="submit"
    >
      Verify Code
    </button>
    <details
      class="govuk-details"
      data-module="govuk-details"
    >
      <summary
        class="govuk-details__summary"
        data-test="problem-with-code"
      >
        <span
          class="govuk-details__summary-text"
        >
          Problem with the code?
        </span>
      </summary>
      <div
        class="govuk-details__text"
      >
        <p>
          We can 
          <a
            class="govuk-link"
            data-test="send-code-again-link"
            href="/send-code-again?notYou=false&action=sendCodeAgain"
          >
            send the code again
          </a>
           or you can 
          <a
            class="govuk-link"
            data-test="not-you-link"
            href="/send-code-again?notYou=true"
          >
            use a different email address.
          </a>
        </p>
        <p>
          If you don't know your email address, contact the member of your team responsible for managing Bichard7 accounts.
        </p>
      </div>
    </details>
  </form>
</div>
`;
